lib/promoted/ruby/client.rb

```diff
@@ -5,67 +5,103 @@
+          throw :invalid_shadow_traffic_delivery_percent if @shadow_traffic_delivery_percent < 0 || @shadow_traffic_delivery_percent > 1.0
+
+          @sampler = Sampler.new
+        end
+        
+        def send_request payload, endpoint=nil
+          endpoint ||= BASE_URL
+          response = Faraday.post(endpoint) do |req|
+            req.headers['Content-Type'] = 'application/json'
+            req.body                    = payload.to_json
+          end
+          response
         end
-        response
-      end
 
-      def deliver payload={}
-        # TODO
-      end
+        def deliver payload={}
+          # TODO
+        end
 
-      def self.log_request args={}, options={}
-        endpoint = options[:endpoint]
-        payload  = prepare_for_logging(args)
-        send_request(payload, endpoint)
-      end
+        def perform_checks?
+          @perform_checks
+        end
 
-      def self.prepare_for_logging args
-        options.set_request_params(args)
-        if options.perform_checks
-          Promoted::Ruby::Client::Settings.check_that_log_ids_not_set(args)
-          pre_delivery_fillin_fields
+        def log_request args={}, options={}
+          endpoint = options[:endpoint]
+          payload  = prepare_for_logging(args)
+          send_request(payload, endpoint)
         end
-        options.log_request_params
-      end
 
-      def self.pre_delivery_fillin_fields
-        if !options.timing[:client_log_timestamp].present?
-          options.client_log_timestamp = Time.now.to_i
+        def should_send_as_shadow_traffic?
+          @sampler.sample_random?(@shadow_traffic_delivery_percent)
         end
-      end
 
-      def self.options
-        @options ||= Options.new
-      end
+        def prepare_for_logging args
+          args = Promoted::Ruby::Client::Util.translate_args(args)
+
+          log_request_builder = LogRequestBuilder.new({
+            only_log: @only_log,
+            uuid: @uuid,
+            now_millis: @now_millis,
+            delivery_timeout_millis: @delivery_timeout_millis,
+            metrics_timeout_millis: @metrics_timeout_millis,
+            should_apply_treatment: @should_apply_treatment
+          })
 
-      def self.promoted_client_impl  params={}
-        # Dummy implementation
-        # TODO will implement it in details
-        perform_checks          = params[:perform_checks] || true
-        only_Log                = params[:only_Log] || false
-        uuid                    = params[:uuid]
-        now_millis              = params[:now_millis] || Time.now.to_i
-        delivery_timeout_millis = params[:delivery_timeout_millis] || DEFAULT_DELIVERY_TIMEOUT_MILLIS
-        metrics_timeout_millis  = params[:metrics_timeout_millis] || DEFAULT_METRICS_TIMEOUT_MILLIS
-        should_apply_treatment  = params[:should_apply_treatment] || false
+          # Note: This method expects as JSON (string keys) but internally, LogRequestBuilder
```

Comment:
If most clients use objects with Symbols, feel free to change it.

I was confused by the use of Symbols vs String keys in the previous PRs.  Hammad and I chatted a couple times about it.  I said to do whatever most clients use and also look at latency.

Reply:
I went back and forth on this a bit, but I think the way it is now, it's slightly easier to test, so I'll leave it for now.